---
title: "Differential Gene Expression of Developmental stages in *Botrylus schlosseri*"
author: "Celeste Valdivia"
output: md_document
date: "2023-04-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, warning = FALSE, message = FALSE)
```
This is file 3 of 3.

## 1. Accessing reference genome
For this workflow, we will utilize the assembled genome of [*Botryllus schlosseri*](https://elifesciences.org/articles/00569) as a reference when quantifying the transcripts of our transcriptome samples.

This code chunk is 'curling' the zip of the *Botryllus schlosseri* genome assembly. 
```{r, engine= 'bash'}
cd /home/shared/8TB_HDD_01/cvaldi/celeste-tunicate-devo/data
curl --insecure -O https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/444/245/GCA_000444245.1_356a-chromosome-assembly/GCA_000444245.1_356a-chromosome-assembly_genomic.fna.gz
```

To unzip assembly and make sure you are in the data directory. Copy and paste the following into your terminal.

```{r, engine= 'bash'}
gzip -d GCA_000444245.1_356a-chromosome-assembly_genomic.fna.gz
```


## 2. Locate sequence reads
You should have used file 1 to download the sequence reads in a large harddrive. For this project the path for our sequence read fastq files are in:

```{r, engine= 'bash'}
# /home/shared/8TB_HDD_01/cvaldi/tunicate-fish546/raw
```

## 3. Build an index file
We will be aligning our sequence reads to the reference genome

The below command takes the path to the HISAT2 executable.
Then with -f option we specify that the input is a fasta file (which can come in the fasta, fna, or fa file extension format).

We then title and create an index which is stored into the output folder. If you don't have an output folder in your working directory already, add one before running the below code.

```{r, engine= 'bash'}
/home/shared/hisat2-2.2.1/hisat2-build \
-f ../data/GCA_000444245.1_356a-chromosome-assembly_genomic.fna \
../output/GCA_000444245.1_356a-valid.index
```

## 3. Perform alignment
Here we will align paired-end RNAseq files against the index we created using the HISAT2 software.

